tests/Regression.hs

```diff
@@ -20,94 +22,82 @@ main = defaultMain tests
-issue97 diff = testCase "issue-97" $ assertNear 5000 $ diff f 0 where f = sum . replicate 5000
+issue97 diff = testCase "issue-97" $ expect eq 5000 $ snd $ diff f 0 where f = sum . replicate 5000
 
 issue104 :: Diff -> Grad -> TestTree
 issue104 diff grad = testGroup "issue-104" [inside, outside] where
   inside = testGroup "inside" [tdiff, tgrad] where
     tdiff = testCase "diff" $ do
-      assertNearList [nan, nan] $ diff (0 `f`) <$> [0, 1]
-      assertNearList [inf, 0.5] $ diff (1 `f`) <$> [0, 1]
-      assertNearList [nan, nan] $ diff (`f` 0) <$> [0, 1]
-      assertNearList [inf, 0.5] $ diff (`f` 1) <$> [0, 1]
+      expect (list eq) [nan, nan] $ snd . diff (0 `f`) <$> [0, 1]
+      expect (list eq) [inf, 0.5] $ snd . diff (1 `f`) <$> [0, 1]
+      expect (list eq) [nan, nan] $ snd . diff (`f` 0) <$> [0, 1]
+      expect (list eq) [inf, 0.5] $ snd . diff (`f` 1) <$> [0, 1]
     tgrad = testCase "grad" $ do
-      assertNearList [nan, nan] $ grad (binary f) [0, 0]
-      assertNearList [nan, inf] $ grad (binary f) [1, 0]
-      assertNearList [inf, nan] $ grad (binary f) [0, 1]
-      assertNearList [0.5, 0.5] $ grad (binary f) [1, 1]
+      expect (list eq) [nan, nan] $ grad (binary f) [0, 0]
+      expect (list eq) [nan, inf] $ grad (binary f) [1, 0]
+      expect (list eq) [inf, nan] $ grad (binary f) [0, 1]
+      expect (list eq) [0.5, 0.5] $ grad (binary f) [1, 1]
     f x y = sqrt $ x * y -- grad f [x, y] = [y / (2 * f x y), x / (2 * f x y)]
   outside = testGroup "outside" [tdiff, tgrad] where
     tdiff = testCase "diff" $ do
-      assertNearList [nan, 0.0] $ diff (0 `f`) <$> [0, 1]
-      assertNearList [inf, 0.5] $ diff (1 `f`) <$> [0, 1]
-      assertNearList [nan, 0.0] $ diff (`f` 0) <$> [0, 1]
-      assertNearList [inf, 0.5] $ diff (`f` 1) <$> [0, 1]
+      expect (list eq) [nan, 0.0] $ snd . diff (0 `f`) <$> [0, 1]
+      expect (list eq) [inf, 0.5] $ snd . diff (1 `f`) <$> [0, 1]
+      expect (list eq) [nan, 0.0] $ snd . diff (`f` 0) <$> [0, 1]
+      expect (list eq) [inf, 0.5] $ snd . diff (`f` 1) <$> [0, 1]
     tgrad = testCase "grad" $ do
-      assertNearList [nan, nan] $ grad (binary f) [0, 0]
-      assertNearList [0.0, inf] $ grad (binary f) [1, 0]
-      assertNearList [inf, 0.0] $ grad (binary f) [0, 1]
-      assertNearList [0.5, 0.5] $ grad (binary f) [1, 1]
+      expect (list eq) [nan, nan] $ grad (binary f) [0, 0]
+      expect (list eq) [0.0, inf] $ grad (binary f) [1, 0]
+      expect (list eq) [inf, 0.0] $ grad (binary f) [0, 1]
+      expect (list eq) [0.5, 0.5] $ grad (binary f) [1, 1]
     f x y = sqrt x * sqrt y -- grad f [x, y] = [sqrt y / 2 sqrt x, sqrt x / 2 sqrt y]
-  binary f = \ [x, y] -> f x y
+  binary f [x, y] = f x y
 
-near :: Double -> Double -> Bool
-near a b = bothNaN || bothInfinite || abs (a - b) <= 1e-12 where
-  bothNaN = isNaN a && isNaN b
-  bothInfinite = signum a == signum b && isInfinite a && isInfinite b
+eq :: Double -> Double -> Bool
+eq a b = isNaN a && isNaN b || a == b
```

Comment:
Also leave a reference to the issue that explains why the implementation of `eq` can't be more precise than this.